### PR TITLE
Replacing RxJava version 1 with 2

### PR DESCRIPTION
**Pull request by tomaszpolanski**

### CODE REVIEW

File: app/src/main/java/com/futurice/freesound/feature/search/DefaultSearchDataModel.java [hunk 1]
```diff
@@ -37,33 +38,33 @@
...
     @NonNull
-    public Observable<Unit> querySearch(@NonNull final String query) {
+    public Single<Unit> querySearch(@NonNull final String query) {
         return freeSoundSearchService.search(get(query))
                                      .map(SoundSearchResult::results)
                                      .map(Option::ofObj)
-                                     .doOnNext(lastResults)
-                                     .map(Unit::asUnit);
+                                     .doOnSuccess(lastResults::onNext)
+                                     .map(it -> Unit.DEFAULT);
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/DefaultSearchDataModel.java]: Is `Unit::asUnit` no longer valid? If changing to return `Unit.DEFAULT` directly you could ignore the parameter, so `__ -> Unit.DEFAULT`.

File: app/src/main/java/com/futurice/freesound/feature/search/DefaultSearchDataModel.java [hunk 2]
```diff
@@ -37,33 +38,33 @@
...
     }
 
     @Override
     @NonNull
     public Observable<Option<List<Sound>>> getSearchResults() {
-        return lastResults.asObservable();
+        return lastResults;
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/DefaultSearchDataModel.java]: Ah, so it seems they got rid of `asObservable`. I wonder why.

File: app/src/main/java/com/futurice/freesound/feature/search/DefaultSearchDataModel.java [hunk 3]
```diff
@@ -37,33 +38,33 @@
...
     @Override
     @NonNull
-    public Observable<Unit> clear() {
-        return Observable.just(Option.<List<Sound>>none())
-                         .doOnNext(lastResults)
-                         .map(__ -> Unit.DEFAULT);
+    public Single<Unit> clear() {
+        return Single.just(Option.<List<Sound>>none())
+                     .doOnSuccess(lastResults::onNext)
+                     .map(__ -> Unit.DEFAULT);
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/DefaultSearchDataModel.java]: OK, now we're just being inconsistent with this `.map(__ -> Unit.DEFAULT)` thing 😉.

File: app/src/main/java/com/futurice/freesound/feature/search/DefaultSearchDataModel.java [hunk 4]
```diff
@@ -37,33 +38,33 @@
...
+        return lastResults;
     }
 
     @Override
     @NonNull
-    public Observable<Unit> clear() {
-        return Observable.just(Option.<List<Sound>>none())
-                         .doOnNext(lastResults)
-                         .map(__ -> Unit.DEFAULT);
+    public Single<Unit> clear() {
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/DefaultSearchDataModel.java]: This is probably out of scope for this change, but can you think of a reason why `clear` needs to return an `Observable` or a `Single`? Could we just call `lastResults.onNext(none())` directly and return `void`?

File: app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java [hunk 5]
```diff
@@ -56,7 +58,8 @@
     private final Analytics analytics;
 
     @NonNull
-    private final BehaviorRelay<String> searchTermRelay = BehaviorRelay.create(NO_SEARCH);
+    private final BehaviorSubject<String> searchTermRelay = BehaviorSubject
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java]: If you declare this as a `Subject<String, String>` it might not need the linewrap. :-)

File: app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java [hunk 6]
```diff
@@ -88,23 +90,23 @@ void openSoundDetails(@NonNull final Sound sound) {
...
-                       .switchMap(this::searchOrClear)
+                       .debounce(2, TimeUnit.SECONDS)
+                       .switchMap(it -> searchOrClear(it).toObservable())
                        .observeOn(mainThread())
                        .subscribe(nothing1(),
                                   e -> Timber.e(e, "Error when setting search term"));
     }
 
-    private Observable<Unit> searchOrClear(@NonNull final String s) {
+    private Maybe<Unit> searchOrClear(@NonNull final String s) {
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java]: Revisit this

File: app/src/main/java/com/futurice/freesound/feature/search/SoundItemViewHolder.java [hunk 7]
```diff
@@ -29,11 +29,11 @@
 import android.view.View;
 import android.widget.TextView;
 
-import rx.android.schedulers.AndroidSchedulers;
-import rx.subscriptions.CompositeSubscription;
+import io.reactivex.disposables.CompositeDisposable;
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/SoundItemViewHolder.java]: This file looks like it needs reformatting.

File: app/src/test/java/com/futurice/freesound/feature/search/DefaultSearchDataModelTest.java [hunk 8]
```diff
@@ -73,78 +72,76 @@ public void querySearch_emitsSingleUnit_whenQuerySearchSuccessful() {
         new Arrangement()
                 .withSearchResultsFor(QUERY, dummyResults());
 
-        TestSubscriber<Unit> ts = testSubscribe(defaultSearchDataModel.querySearch(QUERY));
+        TestSubscriber<Unit> ts = defaultSearchDataModel.querySearch(QUERY).test();
 
-        assertThat(ts).hasNoErrors()
-                      .hasReceivedValue(Unit.DEFAULT)
-                      .hasCompleted();
+        ts.assertValue(Unit.DEFAULT);
```
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/feature/search/DefaultSearchDataModelTest.java]: Does the `TestSubscriber` replacement (`.test()`) now automatically assert those other conditions?

Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/feature/search/DefaultSearchDataModelTest.java]: It doesn't (from my quick reading of the document), so this testing less than before.

File: app/src/test/java/com/futurice/freesound/feature/search/DefaultSearchDataModelTest.java [hunk 9]
```diff
@@ -73,78 +72,76 @@ public void querySearch_emitsSingleUnit_whenQuerySearchSuccessful() {
...
     }
 
     @Test
     public void querySearch_returnsResults_whenQuerySearch_withEmptyTerm() {
         final SoundSearchResult result = dummyResults();
         new Arrangement()
                 .withSearchResultsFor("", result);
-        TestSubscriber<Option<List<Sound>>> ts = testSubscribe(
-                defaultSearchDataModel.getSearchResults());
+        TestObserver<Option<List<Sound>>> ts =
```
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/feature/search/DefaultSearchDataModelTest.java]: Why `TestObserver` in this case and `TestSubscriber` in the others?

File: app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.java [hunk 10]
```diff
@@ -69,34 +68,34 @@ public void thumbnailImageUrl_ifNoWaveFormat_returnEmptyString() {
         when(sound.images()).thenReturn(Maps.newHashMap());
         SoundItemViewModel vm = new SoundItemViewModel(sound, navigator);
 
-        TestSubscriber<String> ts = testSubscribe(vm.thumbnailImageUrl());
+        TestObserver<String> ts = vm.thumbnailImageUrl().test();
 
-        assertThat(ts).hasNoErrors()
-                      .hasReceivedValue("");
+        ts.assertNoErrors()
```
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.java]: Formatting

Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/feature/search/SoundItemViewModelTest.java]: (reformat the file)

File: app/src/test/java/com/futurice/freesound/feature/search/DefaultSearchDataModelTest.java (hunk 9 above)
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/feature/search/DefaultSearchDataModelTest.java]: In new version RxJava 1.x, `TestObserver<T>` has been deprecated. I'll have to look into whether it has a future in RxJava2. Have you heard/know anything about that?

File: app/src/test/java/com/futurice/freesound/functional/StringFunctionsTest.java [hunk 11]
```diff
@@ -23,33 +23,33 @@
 public class StringFunctionsTest {
 
     @Test
-    public void isEmpty_returnsTrue_whenCalled_withEmptyString() {
-        assertThat(StringFunctions.isEmpty().call("")).isTrue();
+    public void isEmpty_returnsTrue_whenCalled_withEmptyString() throws Exception {
+        assertThat(StringFunctions.isEmpty().test("")).isTrue();
```
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/functional/StringFunctionsTest.java]: Is`test` now the `call` for `Predicate`? I'd like to think that API will be renamed, it's a little confusing when compared with the use of `test` on `Observable`.

File: app/src/test/java/com/futurice/freesound/test/rx/IgnoringSingleObserver.java [hunk 12]
```diff
@@ -14,25 +14,32 @@
...
 
-import rx.Observable;
-import rx.observers.TestSubscriber;
+import io.reactivex.SingleObserver;
+import io.reactivex.disposables.Disposable;
 
-import static com.futurice.freesound.utils.Preconditions.get;
-
-public final class TestSubscriberUtils {
+public class IgnoringSingleObserver<T> implements SingleObserver<T> {
```
Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/test/rx/IgnoringSingleObserver.java]: `final`

Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/test/rx/IgnoringSingleObserver.java]: This should also have a private constructor.

**Comment by peter-tackage** [comment]
Interesting stuff. I've only quickly gone through this without pulling the branch. There's a handful of small revisions needed. I'll need to confirm the `TestSubscriber` vs `TestObserver` usage too.

File: app/src/main/java/com/futurice/freesound/feature/search/DefaultSearchDataModel.java (hunk 2 above)
Comment (tomaszpolanski) [inline app/src/main/java/com/futurice/freesound/feature/search/DefaultSearchDataModel.java]: They've just rename it to `hide`, I'll fix it

File: app/src/main/java/com/futurice/freesound/feature/search/DefaultSearchDataModel.java (hunk 4 above)
Comment (tomaszpolanski) [inline app/src/main/java/com/futurice/freesound/feature/search/DefaultSearchDataModel.java]: Could be, but not part of this PR ^^

File: app/src/test/java/com/futurice/freesound/functional/StringFunctionsTest.java (hunk 11 above)
Comment (tomaszpolanski) [inline app/src/test/java/com/futurice/freesound/functional/StringFunctionsTest.java]: It is renamed, as predicate tests stuff. So I dont think we should to rename it

Comment (peter-tackage) [inline app/src/test/java/com/futurice/freesound/functional/StringFunctionsTest.java]: No, we can't rename it, so no need to change it. I was just speculating if they might (or perhaps they will change the `test` method on `Observable`.

File: app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java [hunk 13]
```diff
@@ -88,23 +90,23 @@ void openSoundDetails(@NonNull final Sound sound) {
...
 
     @Override
-    public void bind(@NonNull final CompositeSubscription subscriptions) {
+    public void bind(@NonNull final CompositeDisposable disposables) {
         searchTermRelay.subscribeOn(Schedulers.computation())
                        .observeOn(Schedulers.computation())
                        .map(String::trim)
-                       .switchMap(this::searchOrClear)
+                       .debounce(2, TimeUnit.SECONDS)
+                       .switchMap(it -> searchOrClear(it).toObservable())
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java]: Make `it -> searchOrClear(it).toObservable()` a named method which exposes the `Observable`, rather than the `Maybe`.

File: app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java (hunk 6 above)
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java]: Give the parameter a more meaningful name.

File: app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java [hunk 14]
```diff
@@ -88,23 +90,23 @@ void openSoundDetails(@NonNull final Sound sound) {
...
         return TextUtils.isNullOrEmpty(s) ?
-                searchDataModel.clear() :
-                Observable.just(s)
-                          .filter(StringFunctions.isNotEmpty())
-                          .debounce(2, TimeUnit.SECONDS)
-                          .switchMap(searchDataModel::querySearch);
+                searchDataModel.clear().toMaybe() :
+                Single.just(s)
+                      .filter(StringFunctions.isNotEmpty())
+                      .flatMap(it -> searchDataModel.querySearch(it).toMaybe());
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java]: This used to be a `switchMap` - any reason to change to a `flatmap`? Actually, given that it's a `Single` it shouldn't make a difference in terms of output, only performance if `switchMap` is faster than `flatmap` for example.

File: app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java (hunk 6 above)
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java]: `@NonNull`

File: app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java (hunk 5 above)
Comment (tomaszpolanski) [inline app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java]: Can change that, the good thing is now subject is just `Subject<String>`

File: app/src/main/java/com/futurice/freesound/feature/search/SoundItemViewHolder.java (hunk 7 above)
Comment (tomaszpolanski) [inline app/src/main/java/com/futurice/freesound/feature/search/SoundItemViewHolder.java]: Not really, the formatting doesn't change anything

Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/SoundItemViewHolder.java]: Have you set the formatter for the project? I ran a reformat then and it moved some lines around. Use the formatter settings in the repo: https://github.com/futurice/freesound-android/tree/master/ide/android-studio/code-style

File: app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java (hunk 14 above)
Comment (tomaszpolanski) [inline app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java]: The reason is that flatMap is fastest from all the mapping operators

File: app/src/main/java/com/futurice/freesound/feature/home/HomeViewModel.java [hunk 15]
```diff
@@ -18,38 +18,40 @@
...
-        openSearch.call(Unit.DEFAULT);
+        openSearch.onNext(Unit.DEFAULT);
     }
 
     @Override
-    public void bind(@NonNull final CompositeSubscription subscriptions) {
-        openSearch.subscribeOn(AndroidSchedulers.mainThread())
-                  .observeOn(Schedulers.computation())
+    public void bind(@NonNull final CompositeDisposable disposables) {
+        openSearch.subscribeOn(mainThread())
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/home/HomeViewModel.java]: Why is the `disposables` unused here - shouldn't the Disposable created be added to it?

File: app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java [hunk 16]
```diff
@@ -88,23 +95,29 @@ void openSoundDetails(@NonNull final Sound sound) {
     }
 
     @Override
-    public void bind(@NonNull final CompositeSubscription subscriptions) {
+    public void bind(@NonNull final CompositeDisposable disposables) {
         searchTermRelay.subscribeOn(Schedulers.computation())
```
Comment (peter-tackage) [inline app/src/main/java/com/futurice/freesound/feature/search/SearchViewModel.java]: Add `Disposable` to `disposables`

**Comment by peter-tackage** [comment]
@tomaszpolanski I've two more review comments about making sure that Disposables are added to the Composite in bind().
